File: iosApp/iosApp/Components/ArticleCard.swift
import ComposeApp
import SwiftUI

/// The bordered card Home's Latest section is built from: the same sourcechip, title and
/// mono meta line as a row, with room between them for what the piece says.
///
/// Every card is the same height closed, whatever the length of its title or text — a
/// column of cards that each stop somewhere different reads as a mistake rather than as
/// variety. Room the text does not use is left empty, and text that does not fit is
/// behind "more".
struct ArticleCard: View {
    let host: String
    let title: String
    /// The article's own opening. Nothing on this platform summarises, so it stays the
    /// author's words; see `UnavailableSummariser`.
    let text: String
    var timeAgo: String?
    var meta: [RowMetaItem] = []
    /// Recession, not a strikethrough — the same thing `RowTone.faded` means on a row.
    var faded: Bool = false
    var onTap: () -> Void = {}

    @Environment(\.dusk) private var dusk
    @State private var expanded = false
    @State private var shownHeight: CGFloat = 0
    @State private var fullHeight: CGFloat = 0

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 10) {
                MonogramBadge(host: host)
                Text(host)
                    .dusk(.code)
                    .foregroundStyle(dusk.onSurfaceVariant)
                    .lineLimit(1)
                Spacer(minLength: 8)
                if let timeAgo {
                    Text(timeAgo)
                        .dusk(.code)
                        .foregroundStyle(dusk.onSurfaceVariant)
                }
            }

            Color.clear.frame(height: 14)

            Text(title)
                .dusk(.titleMedium)
                .foregroundStyle(dusk.onSurface)
                // Both, and equal: the second line is held open for a one-line title.
                .lineLimit(Self.titleLines, reservesSpace: true)
                .multilineTextAlignment(.leading)
                .fixedSize(horizontal: false, vertical: true)

            Color.clear.frame(height: 9)

            bodyText
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)
                .measured { shownHeight = $0 }
                // SwiftUI has no `hasVisualOverflow`, so "does this fit" is answered by
                // laying the same text out a second time with nothing holding it back.
                .background(alignment: .topLeading) { probe }

            Color.clear.frame(height: 14)

            HStack(spacing: 10) {
                ForEach(meta) { item in
                    Text(item.text)
                        .dusk(.code)
                        .foregroundStyle(item.accent ? dusk.primary : dusk.onSurfaceVariant)
                        .lineLimit(1)
                }
                Spacer(minLength: 8)

                // Only offered when there is something behind it. A "more" that opens two
                // more words is a broken promise.
                if truncated {
                    Button {
                        withAnimation(Motion.ease(Motion.chip)) { expanded.toggle() }
                    } label: {
                        Text(expanded ? "LESS" : "MORE")
                            .dusk(.code)
                            .foregroundStyle(dusk.onSurfaceVariant)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(Self.cardPadding)
        .frame(maxWidth: .infinity, alignment: .leading)
        .overlay(
            RoundedRectangle(cornerRadius: Radius.card)
                .stroke(dusk.outlineVariant, lineWidth: Stroke.hairline)
        )
        .opacity(faded ? 0.5 : 1)
        .contentShape(Rectangle())
        // The card's own tap opens the article; "more" is a button inside it and takes
        // its own tap first.
        .onTapGesture(perform: onTap)
    }

    /// Two branches rather than one call: `reservesSpace` is what holds every closed
    /// card to the same height, and it has no meaning once the card is open.
    @ViewBuilder
    private var bodyText: some View {
        let styled = Text(text)
            .dusk(.bodyMedium)
            .foregroundStyle(dusk.onSurfaceVariant)
            .multilineTextAlignment(.leading)

        if expanded {
            styled.lineLimit(nil)
        } else {
            styled.lineLimit(Self.bodyLines, reservesSpace: true)
        }
    }

    /// The same text unbounded and invisible, only ever measured.
    private var probe: some View {
        Text(text)
            .dusk(.bodyMedium)
            .fixedSize(horizontal: false, vertical: true)
            .hidden()
            .measured { fullHeight = $0 }
    }

    /// Asked of the closed card only: open, nothing overflows, and reading it then would
    /// take "less" away the moment it was needed.
    private var truncated: Bool { expanded || fullHeight > shownHeight + 1 }

    /// Generous next to a list row's, because the card's whole point is the room.
    private static let cardPadding: CGFloat = 16

    /// Held open whether the title needs both or not; see the note on uniform height.
    private static let titleLines = 2

    private static let bodyLines = 4
}

private extension View {
    func measured(_ onChange: @escaping (CGFloat) -> Void) -> some View {
        overlay(
            GeometryReader { geometry in
                Color.clear
                    .onAppear { onChange(geometry.size.height) }
                    .onChange(of: geometry.size.height) { _, height in onChange(height) }
            }
        )
    }
}

File: composeApp/src/commonMain/kotlin/dev/mks/duskread/ui/common/ArticleCard.kt
package dev.mks.duskread.ui.common

import androidx.compose.animation.Crossfade
import androidx.compose.animation.animateContentSize
import androidx.compose.animation.core.tween
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.mks.duskread.ui.theme.Mono
import dev.mks.duskread.ui.theme.Motion
import dev.mks.duskread.ui.theme.Radius
import dev.mks.duskread.ui.theme.Stroke

/**
 * The bordered card Home's Latest section is built from: the same sourcechip, title and
 * mono meta line as [ListRow], given room to also say what the piece is about.
 *
 * Every card is the same height closed, whatever the length of its title or body — a
 * column of cards that each stop somewhere different reads as a mistake rather than as
 * variety. Room the text does not use is left empty, and text that does not fit is
 * behind "more".
 */
@Composable
fun ArticleCard(
    host: String,
    title: String,
    /** The article's own opening, or a summary of it. Crossfaded, because it can change under the reader. */
    body: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    /** "2d ago", set apart from [meta] because it is the one fact the card is sorted by. */
    timeAgo: String? = null,
    /** Recession, not a strikethrough — the same thing [RowTone.Faded] means on a row. */
    faded: Boolean = false,
    meta: @Composable RowScope.() -> Unit,
) {
    val scheme = MaterialTheme.colorScheme

    // Reset by the body it describes: a summary arriving where an excerpt was is a
    // different piece of text, and it has not been opened.
    var expanded by remember(body) { mutableStateOf(false) }

    // Only ever set while closed. Open, nothing overflows, and reading the flag then
    // would take "less" away the moment it was needed.
    var truncated by remember(body) { mutableStateOf(false) }

    Column(
        modifier
            .fillMaxWidth()
            .alpha(if (faded) 0.5f else 1f)
            .clip(RoundedCornerShape(Radius.Card))
            .border(Stroke.Hairline, scheme.outlineVariant, RoundedCornerShape(Radius.Card))
            .clickable(onClick = onClick)
            // The card grows under the reader's own finger, so the growth is shown.
            .animateContentSize(tween(Motion.Chip))
            .padding(CardPadding),
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            MonogramBadge(host = host, size = ChipSize)
            Spacer(Modifier.width(10.dp))
            RowMeta(host, modifier = Modifier.weight(1f))
            timeAgo?.let { RowMeta(it) }
        }

        Spacer(Modifier.height(14.dp))
        Text(
            text = title,
            style = MaterialTheme.typography.titleMedium,
            fontSize = 17.sp,
            lineHeight = 23.sp,
            // Both, and equal: the second line is held open for a one-line title.
            minLines = TitleLines,
            maxLines = TitleLines,
            overflow = TextOverflow.Ellipsis,
            color = scheme.onSurface,
        )

        Spacer(Modifier.height(9.dp))
        // Keyed on the text itself: an excerpt replaced by a summary changed without the
        // reader asking, and a cut explains that better than a swap.
        Crossfade(targetState = body, animationSpec = tween(Motion.Fade), label = "card-body") { text ->
            Text(
                text = text,
                style = MaterialTheme.typography.bodyMedium,
                fontSize = 13.5.sp,
                lineHeight = 20.sp,
                minLines = BodyLines,
                maxLines = if (expanded) Int.MAX_VALUE else BodyLines,
                overflow = TextOverflow.Ellipsis,
                color = scheme.onSurfaceVariant,
                onTextLayout = { if (!expanded) truncated = it.hasVisualOverflow },
            )
        }

        Spacer(Modifier.height(14.dp))
        Row(verticalAlignment = Alignment.CenterVertically) {
            Row(Modifier.weight(1f), horizontalArrangement = Arrangement.spacedBy(10.dp), content = meta)

            // Only offered when there is something behind it. A "more" that opens two
            // more words is a broken promise.
            if (truncated) {
                Text(
                    text = if (expanded) "LESS" else "MORE",
                    fontFamily = Mono,
                    fontSize = 10.5.sp,
                    color = scheme.onSurfaceVariant,
                    // Its own clickable, so opening the card's text is not opening the
                    // article — the inner one takes the tap.
                    modifier = Modifier
                        .clip(RoundedCornerShape(Radius.Chip))
                        .clickable { expanded = !expanded }
                        .padding(horizontal = 8.dp, vertical = 4.dp),
                )
            }
        }
    }
}

/** Generous next to a list row's, because the card's whole point is the room. */
private val CardPadding = 16.dp

/** Held open whether the title needs both or not; see the note on uniform height above. */
private const val TitleLines = 2

/** Four lines is about what [excerptOf]'s budget produces on a phone. */
private const val BodyLines = 4
